Six tests fail on main (CLI validation messages swallowed, rq-dashboard 500 on non-JSON job meta, regressor splits broken on pandas >= 2.2)

## Summary

Six tests fail on a clean checkout of `main` (currently `60cfda818`). They are unrelated to each other, but each one is masking a **real defect**, not just a stale assertion — so they are worth fixing rather than skipping.

```
FAILED flexmeasures/cli/tests/test_monitor.py::test_monitor_last_seen_rejects_unknown_account_id
FAILED flexmeasures/ui/tests/test_tasks_views.py::test_job_detail_page_loads_with_non_json_metadata
FAILED flexmeasures/data/tests/test_forecasting_pipeline.py::test_future_regressor_split_selects_latest_known_value_per_regressor
FAILED flexmeasures/data/tests/test_forecasting_pipeline.py::test_past_regressor_split_selects_latest_known_value_per_regressor
FAILED flexmeasures/data/tests/test_forecasting_pipeline.py::test_future_regressor_splits_use_only_beliefs_known_at_forecast_belief_time
FAILED flexmeasures/data/tests/test_forecasting_pipeline.py::test_realized_future_regressors_use_latest_known_per_regressor_per_step
```

Reproduce:

```bash
git checkout main
pytest flexmeasures/cli/tests/test_monitor.py::test_monitor_last_seen_rejects_unknown_account_id \
       flexmeasures/ui/tests/test_tasks_views.py::test_job_detail_page_loads_with_non_json_metadata \
       flexmeasures/data/tests/test_forecasting_pipeline.py -q
```

---

### 1. All CLI validation messages are swallowed (`test_monitor_last_seen_rejects_unknown_account_id`)

The test expects `flexmeasures monitor last-seen --account 9999` to say *"No account found with id 9999"*. It actually says:

```
Error: Invalid value for '--account': 9999
```

The reason is not specific to `monitor` — **it affects every CLI option typed with a marshmallow field** (`--account`, `--asset`, `--sensor`, `--source`, …), so users never learn *why* their input was rejected.

Root cause: PR #2219 (Click ≥8.2 upgrade) removed `click.ParamType` from `MarshmallowClickMixin`'s bases (to resolve a Python 3.10 MRO conflict) and added a `__call__` fallback. Click therefore wraps these fields in `FuncParamType`, whose `convert()` does:

```python
try:
    return self.func(value)
except ValueError:
    self.fail(value, param, ctx)   # <- message is the *value*, not the reason
```

`MarshmallowClickMixin.convert()` is now **dead code**, and the `ValueError` raised by `__call__` carries the marshmallow message straight into `self.fail(value, ...)`, which discards it.

### 2. A job with non-JSON meta data 500s the task detail page (`test_job_detail_page_loads_with_non_json_metadata`)

`GET /tasks/0/data/job/<id>.json` returns **500** (`TypeError: Object of type datetime is not JSON serializable`) whenever a job's meta data holds a value the stdlib JSON encoder rejects.

Root cause is upstream: `rq_dashboard/web.py` serializes meta data with a bare `json.dumps(job.get_meta())` — no `default=` (see [rq-dashboard#510](https://github.com/Parallels/rq-dashboard/issues/510), which FlexMeasures already references elsewhere). FlexMeasures works around it today by ISO-stringifying *its own* job meta before enqueueing, but any job that doesn't (a plugin's job, or one whose exception handler stores an exception object) makes the page unreachable.

### 3. Forecasting regressor splits are broken on pandas ≥ 2.2 (4 tests)

`BasePipeline.split_data_all_beliefs()` raises:

```
TypeError: value should be a 'Timestamp', 'NaT', or array of those. Got 'ndarray' instead.
```

from `flexmeasures/data/models/forecasting/pipelines/base.py::_slice_closed`, which does:

```python
a = es.view("int64").to_numpy()          # FutureWarning: Series.view is deprecated
...
out.loc[:, "event_start"] = es.iloc[lo:hi].to_numpy()   # TypeError
```

Two pandas-version problems in four lines: `Series.view` is deprecated in pandas 2.2 (removed in 3.0), and `df.loc[:, col] = <ndarray>` tries to set values *into the existing column's dtype* rather than replacing the column. This is not merely a test-only failure — it breaks the regressor split path itself, i.e. **forecasting with past/future regressors on current pandas**.

## Proposed fixes

1. Raise a click error from `MarshmallowClickMixin.__call__` that carries the marshmallow message, so it survives `FuncParamType`. (Do **not** restore the `click.ParamType` base — that reintroduces the py3.10 MRO conflict PR #2219 fixed.)
2. Patch rq-dashboard's `json.dumps` with a tolerant `default=` in `register_rq_dashboard()`, so non-JSON meta values render instead of 500ing.
3. Drop the deprecated `Series.view("int64")` detour (search directly on `datetime64` values) and replace the column instead of setting into it.

## Notes

* Fixing (1) changes CLI error output from `Invalid value for '--account': 9999` to `Invalid value: No account found with id 9999.` — the option hint is lost (click gives `__call__` no access to the parameter), but the *reason* is restored. Preserving both would require the `ParamType` base, i.e. re-opening the MRO issue.
* (3) is the one with production impact; worth a changelog entry as a bugfix.

## Three more pre-existing failures (data/API suites)

Sweeping `flexmeasures/data/tests/` + `flexmeasures/api/` on clean `main` (`60cfda818`) surfaces three failures beyond the six above — so the total on main is **nine**, not six. PR #2303 fixes these too.

```
FAILED flexmeasures/api/tests/test_task_runs.py::test_api_task_run_get_recent_entry
FAILED flexmeasures/api/common/schemas/tests/test_sensor_data_schema.py::test_build_asset_jobs_data
FAILED flexmeasures/api/v3_0/tests/test_jobs_api.py::test_get_job_status_requires_read_access[test_prosumer_user@seita.nl-200]
```

### 7. `GET /api/ops/getLatestTaskRun` turns any DB error into an opaque 500

`test_api_task_run_get_recent_entry` fails **deterministically**, even in isolation, with a 500. Two distinct bugs:

* **Production bug.** `get_task_run()` (`flexmeasures/api/common/implementations.py`) retries its query on a `DatabaseError` — but does **not roll back** first. A failed statement leaves the transaction aborted, so the retry is rejected with `InFailedSqlTransaction: current transaction is aborted`, masking the original error. The retry can never succeed as written.
* **Test bug.** The test never declared the `setup_api_test_data` fixture, so it ran against a database where `latest_task_run` did not even exist (`UndefinedTable`) — which is what aborted the transaction in the first place.

### 8 + 9. Two tests are order-dependent (Redis job-cache pollution)

Both **pass in isolation** and fail only in a full-suite run, because they assert on job counts / job identity while leftover state from earlier tests sits in the Redis job cache (`@job_cache` also *dedupes* identical scheduling requests, so a cached job from an earlier test is silently reused). Neither test flushed Redis; both now use the existing `clean_redis` fixture.

This is why the failure count varied between identical full-suite runs (7 failed, then 3) before the fix.

--- a/documentation/changelog.rst
+++ b/documentation/changelog.rst
@@ -40,6 +40,7 @@ Bugfixes
 * Fix forecasting with past or future regressors, which raised a ``TypeError`` on pandas 2.2 and higher [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
 * Show why a CLI option was rejected (e.g. "No account found with id 9999") instead of only echoing the offending value [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
 * Fix the task detail page returning a 500 for jobs whose meta data is not JSON-serializable [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
+* Roll back before retrying the query behind ``[GET] /api/ops/getLatestTaskRun``, which otherwise reports an aborted transaction instead of the original database error [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
 * Let storage scheduling treat missing constant SoC bounds as unconstrained lower or upper bounds [see `PR #2221 <https://www.github.com/FlexMeasures/flexmeasures/pull/2221>`_]
 * Allow root assets belonging to different accounts to share the same name, while keeping asset names unique among root assets within the same account and among children of the same parent [see `PR #2226 <https://www.github.com/FlexMeasures/flexmeasures/pull/2226>`_]
 * Fix queued train-predict forecasting jobs losing their resolved forecast window or failing on detached database objects in workers [see `PR #2035 <https://www.github.com/FlexMeasures/flexmeasures/pull/2035>`_]

--- a/documentation/changelog.rst
+++ b/documentation/changelog.rst
@@ -37,6 +37,9 @@ Infrastructure / Support
 
 Bugfixes
 -----------
+* Fix forecasting with past or future regressors, which raised a ``TypeError`` on pandas 2.2 and higher [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
+* Show why a CLI option was rejected (e.g. "No account found with id 9999") instead of only echoing the offending value [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
+* Fix the task detail page returning a 500 for jobs whose meta data is not JSON-serializable [see `PR #2303 <https://www.github.com/FlexMeasures/flexmeasures/pull/2303>`_]
 * Let storage scheduling treat missing constant SoC bounds as unconstrained lower or upper bounds [see `PR #2221 <https://www.github.com/FlexMeasures/flexmeasures/pull/2221>`_]
 * Allow root assets belonging to different accounts to share the same name, while keeping asset names unique among root assets within the same account and among children of the same parent [see `PR #2226 <https://www.github.com/FlexMeasures/flexmeasures/pull/2226>`_]
 * Fix queued train-predict forecasting jobs losing their resolved forecast window or failing on detached database objects in workers [see `PR #2035 <https://www.github.com/FlexMeasures/flexmeasures/pull/2035>`_]

--- a/flexmeasures/api/common/implementations.py
+++ b/flexmeasures/api/common/implementations.py
@@ -62,6 +62,9 @@ def get_task_run():
         ).first()
     except (sqla_exc.ResourceClosedError, sqla_exc.DatabaseError):
         # This is an attempt to make this more stable against some rare condition we encounter. Let's try once more.
+        # NB roll back first: the failed statement leaves the transaction in an aborted state,
+        # in which any further statement (including this retry) is rejected by the database.
+        db.session.rollback()
         time.sleep(2)
         last_known_run = db.session.scalars(
             select(LatestTaskRun).filter(LatestTaskRun.name == task_name).limit(1)

--- a/flexmeasures/api/common/schemas/tests/test_sensor_data_schema.py
+++ b/flexmeasures/api/common/schemas/tests/test_sensor_data_schema.py
@@ -464,7 +464,7 @@ def test_asset_sensors_metadata(
     ]
 
 
-def test_build_asset_jobs_data(db, app, add_battery_assets):
+def test_build_asset_jobs_data(db, app, add_battery_assets, clean_redis):
     """Check that we get both types of jobs for a battery asset."""
     battery_asset = add_battery_assets["Test battery"]
     battery = battery_asset.sensors[0]

--- a/flexmeasures/api/tests/test_task_runs.py
+++ b/flexmeasures/api/tests/test_task_runs.py
@@ -67,7 +67,7 @@ def test_api_task_run_post_no_name(client, requesting_user):
     assert task_run.json["reason"] == "No task name given."
 
 
-def test_api_task_run_get_recent_entry(client):
+def test_api_task_run_get_recent_entry(client, setup_api_test_data):
     task_run = get_task_run(client, "task-B")
     assert task_run.status_code == 200
     assert task_run.json["frequency"] == 10

--- a/flexmeasures/api/v3_0/tests/test_jobs_api.py
+++ b/flexmeasures/api/v3_0/tests/test_jobs_api.py
@@ -54,6 +54,7 @@ def test_get_job_status_requires_read_access(
     add_battery_assets,
     battery_soc_sensor,
     keep_scheduling_queue_empty,
+    clean_redis,
     requesting_user,
     expected_status_code,
 ):

--- a/flexmeasures/data/models/forecasting/pipelines/base.py
+++ b/flexmeasures/data/models/forecasting/pipelines/base.py
@@ -375,18 +375,23 @@ class BasePipeline:
                 if df_ is None or df_.empty:
                     return df_.iloc[0:0].copy() if df_ is not None else None
 
-                # Ensure datetime dtype; then work in int64 ns for searchsorted
+                # Ensure datetime dtype, then search the datetime64 values directly.
+                # (Series.view is deprecated in pandas 2.2, and the int64 detour is unneeded.)
                 es = pd.to_datetime(df_["event_start"], errors="coerce")
-                a = es.view("int64").to_numpy()
+                a = es.to_numpy(dtype="datetime64[ns]")
 
-                lo = np.searchsorted(a, start_ts.value, side="left")
-                hi = np.searchsorted(a, end_ts.value, side="right")  # inclusive end
+                lo = np.searchsorted(a, np.datetime64(start_ts, "ns"), side="left")
+                hi = np.searchsorted(
+                    a, np.datetime64(end_ts, "ns"), side="right"
+                )  # inclusive end
 
                 # Slice original rows by positional indices
                 out = df_.iloc[lo:hi].copy()
                 # (Optional) keep the coerced datetime back on the slice to avoid re-parsing later
                 if not out.empty:
-                    out.loc[:, "event_start"] = es.iloc[lo:hi].to_numpy()
+                    # NB assign the column as a whole, rather than with .loc[:, col],
+                    # which would try to set values into the existing column's dtype.
+                    out["event_start"] = es.iloc[lo:hi].to_numpy()
                 return out
 
             def _latest_known_per_regressor(

--- a/flexmeasures/data/schemas/utils.py
+++ b/flexmeasures/data/schemas/utils.py
@@ -7,6 +7,19 @@ from pint import DefinitionSyntaxError, DimensionalityError, UndefinedUnitError
 from flexmeasures.utils.unit_utils import to_preferred, ur
 
 
+def _format_validation_error(error: ma.exceptions.ValidationError) -> str:
+    """Flatten a marshmallow validation error into a single human-readable line."""
+    messages = error.messages
+    if isinstance(messages, dict):
+        return "; ".join(
+            f"{field}: {' '.join(str(m) for m in msgs) if isinstance(msgs, list) else msgs}"
+            for field, msgs in messages.items()
+        )
+    if isinstance(messages, list):
+        return " ".join(str(message) for message in messages)
+    return str(messages)
+
+
 class MarshmallowClickMixin:
     def __init__(self, *args, **kwargs):
 
@@ -33,11 +46,17 @@ class MarshmallowClickMixin:
             raise click.exceptions.BadParameter(e, ctx=ctx, param=param)
 
     def __call__(self, value):
-        """Support click.FuncParamType by behaving like a conversion callable."""
+        """Support click.FuncParamType by behaving like a conversion callable.
+
+        We raise a click error rather than a ValueError, because click.FuncParamType
+        turns a ValueError into a message that shows only the offending value
+        (`self.fail(value, ...)`), which would swallow the validation message
+        (e.g. "No account found with id 9999.").
+        """
         try:
             return self.deserialize(value)
         except ma.exceptions.ValidationError as e:
-            raise ValueError(e) from e
+            raise click.exceptions.BadParameter(_format_validation_error(e)) from e
 
 
 class FMValidationError(ma.exceptions.ValidationError):

--- a/flexmeasures/ui/__init__.py
+++ b/flexmeasures/ui/__init__.py
@@ -2,7 +2,9 @@
 Backoffice user interface & charting support.
 """
 
+import json
 import os
+from datetime import date, datetime
 
 from flask import current_app, Flask, Blueprint, send_from_directory, request
 from flask_security import login_required, roles_accepted
@@ -83,7 +85,36 @@ def register_at(app: Flask):
     add_jinja_variables(app)
 
 
+def _tolerate_non_json_job_data():
+    """Let rq-dashboard render jobs whose meta data is not JSON-serializable.
+
+    rq-dashboard serializes a job's meta data with a plain ``json.dumps(job.get_meta())``
+    (no ``default=``), so a job carrying e.g. a datetime or an exception in its meta data
+    makes its detail page fail with a 500. We patch the module's ``json`` reference with a
+    shim that falls back to a readable representation instead.
+
+    See https://github.com/Parallels/rq-dashboard/issues/510
+    """
+
+    def default(value):
+        if isinstance(value, (datetime, date)):
+            return value.isoformat()
+        return str(value)
+
+    class TolerantJson:
+        def __getattr__(self, name):
+            return getattr(json, name)
+
+        @staticmethod
+        def dumps(obj, **kwargs):
+            kwargs.setdefault("default", default)
+            return json.dumps(obj, **kwargs)
+
+    rq_dashboard.web.json = TolerantJson()
+
+
 def register_rq_dashboard(app):
+    _tolerate_non_json_job_data()
     app.config.update(
         RQ_DASHBOARD_REDIS_URL=[
             "redis://:%s@%s:%s/%s"

